### ISSUE TITLE
TimeDistributed layer does not (always) propagate mask

### ISSUE THREAD

**Issue by yetanotheryeti** [tensorflow/tensorflow#42127]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.3.0
- Python version: 3.8.2
- CUDA/cuDNN version: CUDA 10.1.243, cuDNN 7.6.5.32

**Describe the current behavior**
The TimeDistributed layer does not propagate a given mask when a custom wrapped layer is used. In the source code (https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/layers/wrappers.py#L192-L245) it shows, that if the "_always_use_reshape" property of the the TimeDistributed layer is false, an RNN-implementation is used, which does not seem to propagate the given mask correctly. "_always_use_reshape" is set to false, if the wrapped layer is a custom one. 
The following simple example shows a custom layer, which simply reduces the last dimension of the input by its sum and multiplies the result by the mask. Note that the mask argument in its call function is not optional (in order to show by a thrown exception that the mask is not propagated).

**Describe the expected behavior**
Running the example code as it is shown in the following, will result in the exception: "_call() missing 1 required positional argument: 'mask'_". This is not expected since a mask is provided in the call of the TimeDistributed layer. However if the "_always_use_reshape" property is set to True, the example works as expected without an exception.

**Standalone code to reproduce the issue**
```
import tensorflow as tf

class TestLayer(tf.keras.layers.Layer):
    def __init__(self):
        super(TestLayer, self).__init__()
        self.supports_masking = True

    def call(self, input, mask):
        # simply sum last dimension
        return tf.reduce_sum(input, axis=-1) * tf.cast(mask, tf.float32)
        
    def compute_output_shape(self, input_shape):
        return input_shape[:-2] + input_shape[-1]
    
    
layer = tf.keras.layers.TimeDistributed(TestLayer())

# By commenting in the following line, this example works properly
# layer._always_use_reshape = True 

result_ones_mask  = layer(tf.random.normal([1,8,4]), mask=tf.ones ([1,8], dtype=tf.bool))
result_zeros_mask = layer(tf.random.normal([1,8,4]), mask=tf.zeros([1,8], dtype=tf.bool))


print(result_ones_mask.numpy())  # expected 1x8-tensor with random numbers
print(result_zeros_mask.numpy()) # expected 1x8-tensor with zeros
```

**Comment by Saduf2019** [comment]
@yetanotheryeti 
I ran the code shared and it seems incomplete please share all dependencies such that we can replicate the issue faced, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/ed4ca48cc24c61b7e1fb301d8582a32a/untitled350.ipynb) for the same. If possible share a colab gist with the issue faced.

**Comment by yetanotheryeti** [comment]
Hi @Saduf2019 

this is the issue I am talking about. Your gist shows exactly the problem. The code (which really is very basic) does not work since the call function requires the mask argument. I thought I made that clear in my description.

Note: As I said, I made the mask argument not optional in purpse to show that it is not set by the TimeDistributed layer. If we set it optional the resulting values in the last code line are not zeros as expected.

**Comment by yetanotheryeti** [comment]
Another follow-up to clarify, that custom layers are problematic with TimeDistributed:
Below a custom layer is constructed, which acts as a "clone" of the GlobalAveragePooling1D layer, by instantiating one internally and only propagating function calls to it.

Switching on the line
`layer = tf.keras.layers.TimeDistributed(ClonedGlobalAveragePooling1D())   # does NOT work`
will result in an error: "Incompatible shape for value ((1, 4)), expected ((1, 1))"

while instead switching on the line
`layer = tf.keras.layers.TimeDistributed(tf.keras.layers.GlobalAveragePooling1D())   # does work`
works as expected without any issue. 

```
import tensorflow as tf

class ClonedGlobalAveragePooling1D(tf.keras.layers.Layer):
    def __init__(self):
        super(ClonedGlobalAveragePooling1D, self).__init__()        
        self.supports_masking = True
        
    def build(self, input_shape):
        self.l = tf.keras.layers.GlobalAveragePooling1D()
        self.l.build(input_shape)

    def call(self, input, mask=None):
        return self.l(input, mask=mask)
    
    def compute_mask(self, inputs, mask=None):
        return self.l.compute_mask(inputs, mask=mask)
        
    def compute_output_shape(self, input_shape):
        return self.l.compute_output_shape(input_shape)
    
    def compute_output_signature(self, input_shape):
        return self.l.compute_output_signature(input_shape)
        

layer = tf.keras.layers.TimeDistributed(ClonedGlobalAveragePooling1D())   # does NOT work
# layer = tf.keras.layers.TimeDistributed(tf.keras.layers.GlobalAveragePooling1D())   # does work

result_ones_mask = layer(tf.random.normal([1,8,4,1]), mask=tf.ones([1,8,4], dtype=tf.bool))
result_zeros_mask = layer(tf.random.normal([1,8,4,1]), mask=tf.zeros([1,8,4], dtype=tf.bool))

print(result_ones_mask.numpy(), result_ones_mask.shape)  # expected 1x8-tensor with random numbers
print(result_zeros_mask.numpy(), result_zeros_mask.shape) # expected 1x8-tensor with zeros
```

**Comment by qlzh727** [comment]
Thanks for filing the issue. Let me clarify some of the question you have:

1. For the https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/layers/wrappers.py#L192-L245, it is propagating the mask correctly since it pass the mask to K.rnn function, which will processing the masking under the hood.

2. In your first example, the mask is created as a required parameter, which makes it incompatible with the TimeDistributed wrapper. By definition, the wrapped layer will see one timestep at a time. It shouldn't need to aware of the mask, since the output value should be auto masked/filtered by the wrapper itself. 

3. In the second example, the mask value you passed is a 3D tensor (1 , 8, 4) which is actually not correct. From the docstring, the mask shape should be a 2D tensor [batch, timestep]. If you change the mask value to tf.zeros([1, 8]), then the code will run correctly.

4. If I change the second example to run with tf.keras.layers.GlobalAveragePooling1D() and mask shape [1, 8], it will actually fail, which is a bug in the current code. I will address that in a change later.

**Comment by yetanotheryeti** [comment]
Hey, qlzh727!

### 1.

I see. So the behavior then differs from the non-rnn version:
- rnn-version: mask gets propagated implicitly by the inner rnn function.
- non-rnn-version: mask gets propagated to the wrapped layer. 

Therefore the input for the wrapped layer also varies depending on the version. See code excursion at the end of this post.

This makes sense, but is still interesting to know and to keep in mind (at least for me). Nevertheless, the argument description for the mask is then confusing, in my opinion, as it says that the mask is propagated to the wrapped layer:
> mask: Binary tensor of shape `(samples, timesteps)` indicating whether a given timestep should be masked. This argument is passed to the wrapped layer (only if the layer supports this argument).

### 2.

> By definition, the wrapped layer will see one timestep at a time. It shouldn't need to aware of the mask, since the output value should be auto masked/filtered by the wrapper itself.

Thanks for the clarification. This matches the preceding observation of the [2 4] input shape in the rnn-version.

### 3.

Oh, yes, my bad.

### 4.

Great! Thanks for looking into the problem and for your effort!

---
Code excursion, clarifying the different behaviours of the rnn- vs non-rnn-version regarding the input shape and mask passed to the wrapped layer:


```
import tensorflow as tf

class TestLayer(tf.keras.layers.Layer):
    def __init__(self):
        super(TestLayer, self).__init__()
        self.supports_masking = True

    def call(self, input, mask=None):
        tf.print("input shape:", tf.shape(input), "; mask:", mask)        
        return input

layer = tf.keras.layers.TimeDistributed(TestLayer())

layer._always_use_reshape = True 

_ = layer(tf.random.normal([2,8,4]), mask=tf.ones ([2,8], dtype=tf.bool))
```
With `layer._always_use_reshape = True` (and thus _non-rnn_) the printed result is: `input shape: [16 4] ; mask: [1 1 1 ... 1 1 1]`
Without it (and thus _rnn_) the printed result is: `input shape: [2 4] ; mask: None` multiple times (one for each time step).

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF v2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/4b0e2c3e68cd434b993a320cd0a81747/untitled101.ipynb)..Thanks !

**Comment by tilakrayal** [comment]
@yetanotheryeti,
I tried to execute the given code with the alternative approach and was able to execute without any issue/error. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/8c6df8fb5336de805b6005fc7434522e/untitled758.ipynb) and let us know if the issue still persists. Thank you!
